### PR TITLE
Done HW!

### PR DESCRIPTION
**Pull request by TkachukTara**

### CODE REVIEW

File: src/main/java/mate/academy/dao/BookDaoImp.java [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+    public List<Book> findAll() {
+        String sql = "select * from books";
+        List<Book> books = new ArrayList<>();
+        try (Connection connection = ConnectionUtil.getConnection();
+                PreparedStatement statement = connection.prepareStatement(sql)) {
+            ResultSet resultSet = statement.executeQuery();
+            while (resultSet.next()) {
+                Long id = resultSet.getLong("id");
+                String title = resultSet.getString("title");
+                BigDecimal price = resultSet.getBigDecimal("price");
```
Comment (sokimaaa) [inline src/main/java/mate/academy/dao/BookDaoImp.java:75]: Repeating code should be extract into separate method. 
```suggestion
                Long id = resultSet.getLong("id");
                String title = resultSet.getString("title");
                BigDecimal price = resultSet.getBigDecimal("price");
```

File: src/main/java/mate/academy/dao/BookDaoImp.java [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+            ResultSet resultSet = statement.executeQuery();
+            if (resultSet.next()) {
+                String title = resultSet.getString("title");
+                BigDecimal price = resultSet.getBigDecimal("price");
+                book = new Book(id, title, price);
+            } else {
+                throw new RuntimeException("Element with this index was not found. Index = " + id);
+            }
+        } catch (SQLException e) {
+            throw new RuntimeException("Could not find element at index.", e);
```
Comment (sokimaaa) [inline src/main/java/mate/academy/dao/BookDaoImp.java:60]: Create and USE custom exception

File: src/main/resources/init_db.sql [hunk 3]
```diff
@@ -0,0 +1,5 @@
+CREATE TABLE books (
+  id INT AUTO_INCREMENT PRIMARY KEY,
```
Comment (sokimaaa) [inline src/main/resources/init_db.sql:2]: When creating a table in MySQL, use bigint column type for storing id.

File: src/main/java/mate/academy/ConnectionUtil.java [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+import java.sql.Connection;
+import java.sql.DriverManager;
+import java.sql.SQLException;
+import java.util.Properties;
+
+public class ConnectionUtil {
+    private static final String DRIVER_CLASS_NAME = "com.mysql.cj.jdbc.Driver";
+    private static final String DB_URL = "jdbc:mysql://localhost:3306/test";
+    private static final Properties dbProperties;
+
```
Comment (sokimaaa) [inline src/main/java/mate/academy/ConnectionUtil.java:12]: util class should have a private constructor to forbid instantiation.

**Review by sokimaaa — CHANGES_REQUESTED** [review]
Go through common mistakes to double check yourself. 

https://github.com/mate-academy/jv-jdbc-intro/blob/main/checklist.md

File: src/main/java/mate/academy/dao/BookDaoImp.java (hunk 2 above)
Comment (sokimaaa) [reply src/main/java/mate/academy/dao/BookDaoImp.java:59]: not fixed

File: src/main/java/mate/academy/dao/BookDaoImp.java [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+            if (resultSet.next()) {
+                book = extractBookFromResultSet(resultSet);
+            } else {
+                throw new DataProcessingException("Element with this index was not found. Index = "
+                        + id, new RuntimeException());
+            }
+        } catch (SQLException e) {
+            throw new RuntimeException("Could not find element at index.", e);
+        }
+        return Optional.of(book);
```
Comment (sokimaaa) [inline src/main/java/mate/academy/dao/BookDaoImp.java:61]: NPE in case of book is null

File: src/main/java/mate/academy/dao/BookDaoImp.java [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+            statement.setBigDecimal(2, book.getPrice());
+            int affectedRows = statement.executeUpdate();
+            if (affectedRows < 1) {
+                throw new DataProcessingException("Expected to update at leas one row,"
+                        + " but updated 0 rows.", new RuntimeException());
+            }
+        } catch (SQLException e) {
+            throw new DataProcessingException("Failed to update element.", e);
+        }
+        return book;
```
Comment (sokimaaa) [inline src/main/java/mate/academy/dao/BookDaoImp.java:96]: Move inside try 
```suggestion
        String sql = "UPDATE books SET title = ?, price = ? WHERE id = ?";
        try (Connection connection = ConnectionUtil.getConnection();
                PreparedStatement statement = connection.prepareStatement(sql)) {
            statement.setLong(3, book.getId());
            statement.setString(1,book.getTitle());
            statement.setBigDecimal(2, book.getPrice());
            int affectedRows = statement.executeUpdate();
            if (affectedRows < 1) {
                throw new DataProcessingException("Expected to update at leas one row,"
                        + " but updated 0 rows.", new RuntimeException());
            }
             return book;
        } catch (SQLException e) {
            throw new DataProcessingException("Failed to update element.", e);
        }
```

**Review by sokimaaa — CHANGES_REQUESTED** [review]

File: src/main/java/mate/academy/ConnectionUtil.java [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+    private ConnectionUtil() {
+        throw new AssertionError("ConnectionUtil class should not be instantiated.");
```
Comment (Sanderfuck) [inline src/main/java/mate/academy/ConnectionUtil.java:14]: How you can achieve this line? It`s redundant

File: src/main/java/mate/academy/Main.java [hunk 8]
```diff
@@ -1,7 +1,22 @@
...
+    public static void main(String[] args) {
+        BookDao bookDao = (BookDao) injector.getInstance(BookDao.class);
+        Book book = new Book("book", new BigDecimal("119.33"));
+        bookDao.create(book);
+        System.out.println(bookDao.findById(10L));
+        for (Book element : bookDao.findAll()) {
+            System.out.println(element);
+        }
+        Book bookUpdate = new Book(5L, "Java", new BigDecimal("999.35"));
+        bookDao.update(bookUpdate);
```
Comment (Sanderfuck) [inline src/main/java/mate/academy/Main.java:20]: Where you delete book?

File: src/main/java/mate/academy/dao/BookDaoImp.java (hunk 5 above)
Comment (Sanderfuck) [reply src/main/java/mate/academy/dao/BookDaoImp.java:61]: Not fixed, you need to check

File: src/main/java/mate/academy/dao/BookDaoImp.java [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+    public boolean deleteById(Long id) {
+        String sql = "DELETE FROM books WHERE id = ?";
+        try (Connection connection = ConnectionUtil.getConnection();
+                PreparedStatement statement = connection.prepareStatement(sql)) {
+            statement.setLong(1, id);
+            int affectedRows = statement.executeUpdate();
+            if (affectedRows < 1) {
+                throw new DataProcessingException("Expected to update at leas one row,"
+                        + " but updated 0 rows.", new RuntimeException());
+            }
```
Comment (Sanderfuck) [inline src/main/java/mate/academy/dao/BookDaoImp.java:109]:
```suggestion
            return affectedRows > 0
```

File: src/main/java/mate/academy/dao/BookDaoImp.java [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+        } catch (SQLException e) {
+            throw new DataProcessingException("Failed to delete element.", e);
+        }
+        return true;
```
Comment (Sanderfuck) [inline src/main/java/mate/academy/dao/BookDaoImp.java:113]:
```suggestion
```

**Review by Sanderfuck — CHANGES_REQUESTED** [review]

File: src/main/java/mate/academy/Main.java (hunk 8 above)
Comment (kerrrusha) [reply src/main/java/mate/academy/Main.java:20]: not fixed. add `bookDao.deleteById(...);` call here

File: src/main/java/mate/academy/dao/BookDaoImp.java [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+            statement.setLong(1, id);
+            ResultSet resultSet = statement.executeQuery();
+            if (resultSet.next()) {
+                book = extractBookFromResultSet(resultSet);
+            } else {
+                throw new DataProcessingException("Element with this index was not found. Index = "
+                        + id, new RuntimeException());
+            }
+        } catch (SQLException e) {
+            throw new NullPointerException();
```
Comment (kerrrusha) [inline src/main/java/mate/academy/dao/BookDaoImp.java:59]:
```suggestion
            throw new DataProcessingException("Couldn't find a book by id " + id, e);
```

File: src/main/java/mate/academy/dao/BookDaoImp.java [hunk 12]
```diff
@@ -0,0 +1,122 @@
...
+        try (Connection connection = ConnectionUtil.getConnection();
+                PreparedStatement statement = connection.prepareStatement(sql)) {
+            statement.setLong(1, id);
+            ResultSet resultSet = statement.executeQuery();
+            if (resultSet.next()) {
+                book = extractBookFromResultSet(resultSet);
+            } else {
+                throw new DataProcessingException("Element with this index was not found. Index = "
+                        + id, new RuntimeException());
+            }
```
Comment (kerrrusha) [inline src/main/java/mate/academy/dao/BookDaoImp.java:57]:
```suggestion
            Book book = null;
            if (resultSet.next()) {
                book = extractBookFromResultSet(resultSet);
            }
            return Optional.ofNullable(book);
```

File: src/main/java/mate/academy/dao/BookDaoImp.java [hunk 13]
```diff
@@ -0,0 +1,122 @@
...
+        } catch (SQLException e) {
+            throw new NullPointerException();
+        }
+        return Optional.of(book);
```
Comment (kerrrusha) [inline src/main/java/mate/academy/dao/BookDaoImp.java:61]:
```suggestion
```

File: src/main/java/mate/academy/dao/BookDaoImp.java (hunk 9 above)
Comment (kerrrusha) [reply src/main/java/mate/academy/dao/BookDaoImp.java:109]: not fixed

File: src/main/java/mate/academy/dao/BookDaoImp.java [hunk 14]
```diff
@@ -0,0 +1,122 @@
...
+        try (Connection connection = ConnectionUtil.getConnection();
+                PreparedStatement statement = connection.prepareStatement(sql)) {
+            statement.setLong(1, id);
+            int affectedRows = statement.executeUpdate();
+            if (affectedRows < 1) {
+                throw new DataProcessingException("Expected to update at leas one row,"
+                        + " but updated 0 rows.", new RuntimeException());
+            }
+        } catch (SQLException e) {
+            throw new DataProcessingException("Failed to delete element.", e);
```
Comment (kerrrusha) [inline src/main/java/mate/academy/dao/BookDaoImp.java:111]:
```suggestion
            throw new DataProcessingException("Failed to delete book by id " + id, e);
```

File: src/main/java/mate/academy/dao/BookDaoImp.java (hunk 10 above)
Comment (kerrrusha) [reply src/main/java/mate/academy/dao/BookDaoImp.java:113]: not fixed

**Review by kerrrusha — CHANGES_REQUESTED** [review]
please fix prev comments before requesting review

**Pull request closed without merging**